src/App.js

```diff
@@ -24,62 +24,87 @@ import getToken from './lib/tokens';
 
 function App() {
   const [listId, setListId] = useState(null);
-  const [userToken, setUserToken] = useState('');
 
-  function saveToken(e) {
-    e.preventDefault();
+  function createList() {
     const token = getToken();
-    localStorage.setItem('token', token);
-    setUserToken(token);
+
+    return db
+      .collection('lists')
+      .add({ token: token })
+      .then((results) => {
+        const { id } = results;
+        if (id === undefined) throw new Error('Failed to create list.');
+        return id;
+      })
+      .then((newListId) => {
+        // list successfully created in db, so update listId state and save token to localStorage
+        setListId(newListId);
+        localStorage.setItem('token', token);
+        return true;
+      });
+  }
+
+  // returns true if token matches a list in db, returns false if no list, throws error on connection problem
+  function isTokenValid(token) {
+    return db
+      .collection('lists')
+      .where('token', '==', token)
+      .get()
+      .then((querySnapshot) => {
+        // if there are results and an id property exists
+        if (!querySnapshot.empty && 'id' in querySnapshot.docs[0]) {
+          setListId(querySnapshot.docs[0].id);
+          return true;
+
+          // check metadata.fromCache to distinguish between no results (invalid token) and a connection issue
+        } else if (!querySnapshot.metadata.fromCache) {
+          return false;
+        } else {
+          throw new Error('Connection problem');
+        }
+      });
+  }
+
+  function joinList(shareToken) {
+    return isTokenValid(shareToken).then((listExists) => {
+      if (listExists) {
+        localStorage.setItem('token', shareToken);
+        return true;
+      } else throw new Error('Invalid token');
+    });
   }
 
   // on component mounting, look for token in local storage and use it to retrieve the list id
   useEffect(() => {
     const token = localStorage.getItem('token');
 
     if (token) {
-      setUserToken(token);
-
-      // find the list in Firestore associated with the stored token from local storage
-      db.collection('lists')
-        .where('token', '==', token)
-        .get()
-        .then((querySnapshot) => {
-          // if there are results and an id property exists
-          if (!querySnapshot.empty && 'id' in querySnapshot.docs[0]) {
-            setListId(querySnapshot.docs[0].id); // save the list id for later
-          }
+      isTokenValid(token)
+        .then((listExists) => {
+          if (listExists !== true) localStorage.removeItem('token');
         })
         .catch((error) => {
           console.log('Error getting list: ', error);
         });
     }
-  }, [listId]);
+  }, []);
 
   return (
     <Router>
       <div className="App container">
         <Switch>
           <Route exact path="/">
-            {userToken ? (
+            {listId ? (
               <Redirect to="/list" />
             ) : (
```

Comment:
Is there a reason why we have changed from checking userToken in the routes to listId?

Reply:
Yep! Do you remember when we built the "Add Item" form, and I mentioned we needed to add a check somewhere to make sure we had a list ID before loading it? Since the item form and shopping list won't work without a list ID (won't know which list to read/write to), we switched the routing logic to check for a list ID instead. You may have noticed we also create a new list in the database when a new token is generated, so a token will always have a list if it is valid.

Reply:
To put it another way, the `listId` is also effectively a check for userToken, because our app will not set a `listId` state without a token.